File: shoring/blackjackEngine/src/test/java/com/panopset/tests/blackjackEngine/BetAmountStrategyTest.java
package com.panopset.tests.blackjackEngine;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.panopset.blackjackEngine.BetAmountStrategy;
import com.panopset.blackjackEngine.BlackjackConfigDefault;
import com.panopset.blackjackEngine.BlackjackGameEngine;

public class BetAmountStrategyTest {

  @Test
  void testDoubleAfterBust() {
    BlackjackGameEngine bge = new BlackjackGameEngine(new BlackjackConfigDefault() {

      @Override
      public boolean isBetIdeaDoubleAfterBust() {
        return true;
      }
    });
    bge.setAutomaticOnForTesting();
    BetAmountStrategy bas = new BetAmountStrategy(bge);
    bge.setPriorHandBustedFlag();
    Assertions.assertEquals(1000, bas.adjust());
    bge.clearPriorHandBustedFlag();
    Assertions.assertEquals(500, bas.adjust());
    bge.clearAutomaticForTesting();
    bge = new BlackjackGameEngine(new BlackjackConfigDefault() {

      @Override
      public boolean isBetIdeaLetItRideAfterTwoWins() {
        return true;
      }

      @Override
      public boolean isCountVeryPositive() {
        return true;
      }

      @Override
      public int getLargeBetInWholeDollars() {
        return 50;
      }
    });
    bge.setAutomaticOnForTesting();
    bas = new BetAmountStrategy(bge);
    Assertions.assertEquals(5000, bas.adjust());
    bge.incrementStreak();
    Assertions.assertEquals(5000, bas.adjust());
    bge.incrementStreak();
    Assertions.assertEquals(10000, bas.adjust());
    bge.clearAutomaticForTesting();
  }
}

File: code/shoring/blackjackEngine/src/test/kotlin/com/panopset/tests/blackjackEngine/BetAmountStrategyTest.kt
package com.panopset.tests.blackjackEngine

import com.panopset.blackjackEngine.BetAmountStrategy
import com.panopset.blackjackEngine.BlackjackConfigDefault
import com.panopset.blackjackEngine.BlackjackGameEngine
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class BetAmountStrategyTest {
    @Test
    fun testDoubleAfterBust() {
        var bge = BlackjackGameEngine(object : BlackjackConfigDefault() {
            override val isBetIdeaDoubleAfterBust: Boolean
                get() = true
        })
        bge.setAutomaticOnForTesting()
        var bas = BetAmountStrategy(bge)
        bge.setPriorHandBustedFlag()
        Assertions.assertEquals(1000, bas.adjust())
        bge.clearPriorHandBustedFlag()
        Assertions.assertEquals(500, bas.adjust())
        bge.clearAutomaticForTesting()
        bge = BlackjackGameEngine(object : BlackjackConfigDefault() {
            override val isBetIdeaLetItRideAfterTwoWins: Boolean
                get() = true
            override val strategicVeryPositiveCount: Int
                get() = Int.MIN_VALUE
            override val largeBetInWholeDollars: Int
                get() = 50
        })
        bge.setAutomaticOnForTesting()
        bas = BetAmountStrategy(bge)
        Assertions.assertEquals(5000, bas.adjust())
        bge.incrementStreak()
        Assertions.assertEquals(5000, bas.adjust())
        bge.incrementStreak()
        Assertions.assertEquals(10000, bas.adjust())
        bge.clearAutomaticForTesting()
    }
}
